fix(*): lower instance priority

fpvandoorn opened the pull request:
use lower instance priority for instances that always apply
also do this for automatically generated instances using the `extend` keyword
also add a comment to most places where we short-circuit type-class inference. This can lead to greatly increased search times (see issue #1561), so we might want to delete some/all of them.

This PR will almost certainly break things, although I strongly expect that it will decrease the overall type-class inference search space. I investigated one place where type-class inference was short-circuited. This short-circuiting was necessary before the change, but not afterwards.

fpvandoorn committed:
fix(*): lower instance priority

use lower instance priority for instances that always apply
also do this for automatically generated instances using the `extend` keyword
also add a comment to most places where we short-circuit type-class inference. This can lead to greatly increased search times (see issue #1561), so we might want to delete some/all of them.

sgouezel commented on src/tactic/lint.lean:
```diff
@@ -266,7 +267,30 @@ meta def instance_priority (d : declaration) : tactic (option string) := do
+  if always_applies then return $ some "set priority below 1000" else return none
+
+/- Note [default priority]:
+  Instances that always apply should be applied after instances that only apply in specific cases.
```
Can you expand a little bit on what you mean by "instances that always apply"? This is not so clear as, of course, no instance can really always apply, you need some extra (typeclass) information. Do you mean that they will be something of the forme `my_class alpha` where `alpha` is some arbitrary type, and then the instance search will go on to check additional properties of `alpha`, in opposition to `my_class (my_nice_constructor alpha)` where it will fail right away if we try to apply it to `my_class beta` where `beta` is not `my_nice_constructor` of something?

fpvandoorn replied:
Yes, this is exactly what I mean. I will clarify it.

fpvandoorn committed:
put default_priority option inside section

Default priority also applies to all instances, not just automatically-generates ones
the scope of set_option is limited to a section

fpvandoorn committed:
two more low priorities

sgouezel commented on src/algebra/ordered_group.lean:
```diff
@@ -6,15 +6,14 @@ Authors: Mario Carneiro, Johannes Hölzl
 
 set_option old_structure_cmd true
-
+set_option default_priority 100 -- see Note [default priority]
```
Missing section?

sgouezel replied:
Or remove?

fpvandoorn replied:
I piggy-backed on `section old_structure_cmd`

fpvandoorn committed:
fix some broken proofs

fpvandoorn committed:
fix proof

fpvandoorn committed:
more fixes

fpvandoorn committed:
more fixes

fpvandoorn committed:
increase max_depth a bit

fpvandoorn committed:
update notes

fpvandoorn committed:
fix linter issues

mergify[bot] committed:
Merge branch 'master' into instance_priority
